bpf/flows.c

```diff
@@ -234,22 +248,98 @@ static inline int fill_ethhdr(struct ethhdr *eth, void *data_end, flow_id *id, u
     return SUBMIT;
 }
 
+static __always_inline void fill_flow_seq_id(flow_seq_id *seq_id, pkt_info *pkt, u32 seq, u8 reversed) {
+    flow_id *id = pkt->id;
+    if (reversed) {
+        __builtin_memcpy(seq_id->src_ip, id->dst_ip, IP_MAX_LEN);
+        __builtin_memcpy(seq_id->dst_ip, id->src_ip, IP_MAX_LEN);
+        seq_id->src_port = id->dst_port;
+        seq_id->dst_port = id->src_port;
+    } else {
+        __builtin_memcpy(seq_id->src_ip, id->src_ip, IP_MAX_LEN);
+        __builtin_memcpy(seq_id->dst_ip, id->dst_ip, IP_MAX_LEN);
+        seq_id->src_port = id->src_port;
+        seq_id->dst_port = id->dst_port;
+    }
+    seq_id->seq_id = seq;
+}
+
+static __always_inline void calculate_flow_rtt(pkt_info *pkt, u8 direction, void *data_end) {
+    flow_seq_id seq_id;
+    __builtin_memset(&seq_id, 0, sizeof(flow_seq_id));
+
+    switch (pkt->id->transport_protocol)
+    {
+    case IPPROTO_TCP: {
+            struct tcphdr *tcp = (struct tcphdr *) pkt->l4_hdr;
+            if ( !tcp || ((void *)tcp + sizeof(*tcp) > data_end) ) {
+                break;
+            }
+            if ((direction == EGRESS) && IS_SYN_PACKET(pkt)) {
```

Comment:
Related to my comment here https://github.com/netobserv/netobserv-ebpf-agent/pull/117#issuecomment-1532669464
I wonder if it would be better to write here:

```c
if (pkt->flags & SYN_FLAG) {
```
and below:

```c
if (pkt->flags & SYN_ACK_FLAG) {
```
(even dropping the direction check)

Since we use a dedicated flag for SYN+ACK ?

Reply:
@jotak referencing the figure from my comment here - https://github.com/netobserv/netobserv-ebpf-agent/issues/102#issuecomment-1520129627

We check against these combinations - 
For Client: SYN (Egress) and SYN+ACK (Ingress)
For Server: SYN+ACK (Egress) and ACK(Ingress)

Hence the checks are like this,
```
#define IS_SYN_PACKET(pkt)    ((pkt->flags & SYN_FLAG) || (pkt->flags & SYN_ACK_FLAG))
#define IS_ACK_PACKET(pkt)    ((pkt->flags & ACK_FLAG) || (pkt->flags & SYN_ACK_FLAG))
```

On the second point, I could potentially drop the direction check but that would mean that a node which sends a SYN packet will also check every ACK that it sends to see if it is present in the `flow_sequences` identifier hashmap (would be an extra check for ACKs originating from that node)